# EE 399 HW 4: Training Neural Nets
## Shaunak Bhide


In [72]:
#import cell
#using keras since it is a lot simpler to work with 
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, LSTM

#### Part I: HW dataset fitting using NN

In [53]:
#Dataset from HW1
X=np.arange(0,31)
Y=np.array([30, 35, 33, 32, 34, 37, 39, 38, 36, 36, 37, 39, 42, 45, 45, 41, 40, 39, 42, 44, 47, 49, 50, 49, 46, 48, 50, 53, 55, 54, 53])

In [55]:
#part i
#creating the model
model = Sequential()
model.add(Dense(16, input_dim=1, activation='relu')) #layer 1, 16 neurons
model.add(Dense(16, activation='relu')) #layer 2, 16 neurons
model.add(Dense(1, activation='linear')) # layer 3, 1 neuron

model.compile(loss='mean_squared_error', optimizer='adam')

In [56]:
# part i continued
#train the model on the entire dataset
model.fit(X, Y, epochs=10000, batch_size=10)


Epoch 1/10000
4/4 [==============================] - 1s 5ms/step - loss: 2134.3005
Epoch 2/10000
4/4 [==============================] - 0s 2ms/step - loss: 2089.1035
Epoch 3/10000
4/4 [==============================] - 0s 2ms/step - loss: 2043.3623
Epoch 4/10000
4/4 [==============================] - 0s 2ms/step - loss: 2002.3040
Epoch 5/10000
4/4 [==============================] - 0s 2ms/step - loss: 1961.8259
Epoch 6/10000
4/4 [==============================] - 0s 3ms/step - loss: 1923.0194
Epoch 7/10000
4/4 [==============================] - 0s 2ms/step - loss: 1883.1189
Epoch 8/10000
4/4 [==============================] - 0s 1ms/step - loss: 1846.1554
Epoch 9/10000
4/4 [==============================] - 0s 2ms/step - loss: 1808.4646
Epoch 10/10000
4/4 [==============================] - 0s 2ms/step - loss: 1770.8621
Epoch 11/10000
4/4 [==============================] - 0s 3ms/step - loss: 1736.1766
Epoch 12/10000
4/4 [==============================] - 0s 2ms/step - loss: 1704.3063
E

In [57]:
#part ii 
#split data: use first 20 points as training and last 10 as test
#training and test data
trainingX = X[:20]
trainingY = Y[:20]
testX = X[-10:]
testY = Y[-10:]

model.fit(trainingX, trainingY, epochs=10000, batch_size=10)

#compute least square error for training data
train_pred = model.predict(trainingX)
train_error = np.mean(np.square(train_pred - trainingY))

#compute least square error for test data
test_pred = model.predict(testX)
test_error = np.mean(np.square(test_pred - testY))




Epoch 1/10000
2/2 [==============================] - 0s 3ms/step - loss: 5.2762
Epoch 2/10000
2/2 [==============================] - 0s 3ms/step - loss: 5.0449
Epoch 3/10000
2/2 [==============================] - 0s 2ms/step - loss: 5.0115
Epoch 4/10000
2/2 [==============================] - 0s 3ms/step - loss: 5.0641
Epoch 5/10000
2/2 [==============================] - 0s 3ms/step - loss: 5.1506
Epoch 6/10000
2/2 [==============================] - 0s 3ms/step - loss: 5.1270
Epoch 7/10000
2/2 [==============================] - 0s 2ms/step - loss: 5.1371
Epoch 8/10000
2/2 [==============================] - 0s 3ms/step - loss: 5.1298
Epoch 9/10000
2/2 [==============================] - 0s 2ms/step - loss: 5.1145
Epoch 10/10000
2/2 [==============================] - 0s 2ms/step - loss: 5.0829
Epoch 11/10000
2/2 [==============================] - 0s 2ms/step - loss: 5.0433
Epoch 12/10000
2/2 [==============================] - 0s 2ms/step - loss: 5.0082
Epoch 13/10000
2/2 [=================

In [58]:
#part iii: using first and last 10 points as training data and middle 10 points as test
#split data
#training and test data
X3 = np.concatenate([X[:10], X[-10:]])
Y3 = np.concatenate([X[:10], X[-10:]])
X3T = X[10:20]
Y3T = Y[10:20]

model.fit(X3, Y3, epochs=10000, batch_size=10)

#compute least square error for training data
train_pred3 = model.predict(X3)
train_error3 = np.mean(np.square(train_pred3 - Y3))

#compute least square error for test data
test_pred3 = model.predict(X3T)
test_error3 = np.mean(np.square(test_pred3 - Y3T))



Epoch 1/10000
2/2 [==============================] - 0s 2ms/step - loss: 681.4082
Epoch 2/10000
2/2 [==============================] - 0s 2ms/step - loss: 536.9318
Epoch 3/10000
2/2 [==============================] - 0s 4ms/step - loss: 405.6796
Epoch 4/10000
2/2 [==============================] - 0s 4ms/step - loss: 314.0229
Epoch 5/10000
2/2 [==============================] - 0s 4ms/step - loss: 271.1825
Epoch 6/10000
2/2 [==============================] - 0s 5ms/step - loss: 258.4582
Epoch 7/10000
2/2 [==============================] - 0s 4ms/step - loss: 266.3463
Epoch 8/10000
2/2 [==============================] - 0s 3ms/step - loss: 273.0779
Epoch 9/10000
2/2 [==============================] - 0s 3ms/step - loss: 270.3575
Epoch 10/10000
2/2 [==============================] - 0s 3ms/step - loss: 257.0956
Epoch 11/10000
2/2 [==============================] - 0s 2ms/step - loss: 238.4334
Epoch 12/10000
2/2 [==============================] - 0s 3ms/step - loss: 217.1878
Epoch 13/1000

In [59]:
print("Training set 1 error = ", train_error)
print("Test set 1 error = ", test_error)

print("Training set 2 error = ", train_error3)
print("Test set 2 error = ", test_error3)

Training set 1 error =  29.525825250436466
Test set 1 error =  26.053422135278232
Training set 2 error =  237.00022363711562
Test set 2 error =  738.4990082935087


### From HW1
#### Training Data Set 1 Loss functions 
- Loss function from line fit  2.242749387090776 
- Loss function from parabola fit  2.2427493872531716 
- Loss function from polynomial fit  0.028351970844281614 

#### Test Data Set 1 Loss functions 

 - Loss function from line fit  3.4392167768613793 
 - Loss function from parabola fit  3.4391905006921237 
 - Loss function from polynomial fit  30022321415.023407

#### Training Data Set 2 Loss functions 

 - Loss function from line fit  3.1615109544610585 
 - Loss function from parabola fit  3.1636135157981684 
 - Loss function from polynomial fit  34471695091300.21 

#### Test Data Set 2 Loss functions 

 - Loss function from line fit  33.573133901279796 
 - Loss function from parabola fit  34.011683144193 
 - Loss function from polynomial fit  4027.9343692981083 

### Neural Net Results: 10000 Epochs, batch size 10

- Training set 1 error =  29.525825250436466
- Test set 1 error =  26.053422135278232
- Training set 2 error =  237.00022363711562
- Test set 2 error =  738.4990082935087




#### Part II MNIST_784 dataset analysis using NN

In [86]:
#imports for part II
#import and preprocess MNIST dataset
from keras.datasets import mnist
from keras.utils import to_categorical

#PCA 
from sklearn.decomposition import PCA

#SVM
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score



In [79]:
# load the MNIST dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# reshape the images to 1D arrays and normalize the pixel values to the range [0, 1]
X_train = X_train.reshape(-1, 784) / 255.
X_test = X_test.reshape(-1, 784) / 255.

# perform PCA analysis to reduce the dimensionality of the data
pca = PCA(n_components=20)
pca.fit(X_train)
X_train_pca = pca.transform(X_train)
X_test_pca = pca.transform(X_test)

# define the neural network architecture
model = Sequential()
model.add(Dense(128, activation='relu', input_shape=(20,)))
model.add(Dense(10, activation='softmax')) 

# compile the model
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# train the model
model.fit(X_train_pca, to_categorical(y_train), validation_data=(X_test_pca, to_categorical(y_test)), epochs=10, batch_size=32)

# evaluate the model
test_loss, test_acc_FF = model.evaluate(X_test_pca, to_categorical(y_test))
test_acc_FF = test_acc_FF * 100
print(f"Test Accuracy (Feed Forward Neural Net): {test_acc_FF:.2f}%")

Epoch 1/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3594 - accuracy: 0.8991 - val_loss: 0.1834 - val_accuracy: 0.9464
Epoch 2/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.1649 - accuracy: 0.9512 - val_loss: 0.1366 - val_accuracy: 0.9594
Epoch 3/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.1303 - accuracy: 0.9607 - val_loss: 0.1179 - val_accuracy: 0.9650
Epoch 4/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.1131 - accuracy: 0.9657 - val_loss: 0.1115 - val_accuracy: 0.9663
Epoch 5/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.1019 - accuracy: 0.9682 - val_loss: 0.1022 - val_accuracy: 0.9691
Epoch 6/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.0943 - accuracy: 0.9705 - val_loss: 0.1019 - val_accuracy: 0.9698
Epoch 7/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.0883 - accuracy: 0.9724 - val_loss: 0.1018 - val_accuracy:

In [80]:
# Reshape the data for the LSTM model
x_train = np.reshape(X_train, (X_train.shape[0], 28, 28))
x_test = np.reshape(X_test, (X_test.shape[0], 28, 28))

#one-hot encoding the labels
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

# Define the LSTM model
model = Sequential()
model.add(LSTM(128, input_shape=(28, 28)))
model.add(Dense(10, activation='softmax'))

# Compile the model
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train, epochs=10, batch_size=128, validation_data=(x_test, y_test))

# Evaluate the model
loss, test_acc_LSTM = model.evaluate(x_test, y_test)
test_acc_LSTM = test_acc_LSTM * 100
print(f"Test Accuracy (LSTM): {test_acc_LSTM:.2f}%")

Epoch 1/10
469/469 [==============================] - 19s 36ms/step - loss: 0.4935 - accuracy: 0.8393 - val_loss: 0.1520 - val_accuracy: 0.9534
Epoch 2/10
469/469 [==============================] - 17s 36ms/step - loss: 0.1452 - accuracy: 0.9562 - val_loss: 0.1182 - val_accuracy: 0.9620
Epoch 3/10
469/469 [==============================] - 17s 35ms/step - loss: 0.0998 - accuracy: 0.9694 - val_loss: 0.0905 - val_accuracy: 0.9733
Epoch 4/10
469/469 [==============================] - 17s 36ms/step - loss: 0.0763 - accuracy: 0.9767 - val_loss: 0.0862 - val_accuracy: 0.9716
Epoch 5/10
469/469 [==============================] - 17s 36ms/step - loss: 0.0632 - accuracy: 0.9805 - val_loss: 0.0683 - val_accuracy: 0.9789
Epoch 6/10
469/469 [==============================] - 17s 36ms/step - loss: 0.0517 - accuracy: 0.9835 - val_loss: 0.0656 - val_accuracy: 0.9789
Epoch 7/10
469/469 [==============================] - 17s 36ms/step - loss: 0.0450 - accuracy: 0.9864 - val_loss: 0.0564 - val_accuracy:

In [87]:
#SVM

(X_train, y_train), (X_test, y_test) = mnist.load_data()

# reshape the images to 1D arrays and normalize the pixel values to the range [0, 1]
X_train = X_train.reshape(-1, 784) / 255.
X_test = X_test.reshape(-1, 784) / 255.



In [88]:
# Train an SVM classifier on the training set
clf_svm = SVC(C=1.0, kernel='rbf', gamma='scale', random_state=42)
clf_svm.fit(X_train, y_train)

# Evaluate the performance of the SVM on the testing set
y_pred_svm = clf_svm.predict(X_test)
accuracy_test_svm = accuracy_score(y_test, y_pred_svm) * 100
print(f"Test Accuracy (SVM): {accuracy_test_svm:.2f}%")

Test Accuracy (SVM): 97.92%
